<h1 style="text-align:  center;">Введение в анализ данных</h1>

<h1 style="text-align:center;">LAB1</h1>

Цель работы: осуществить предварительную обработку данных csv-файла, выявить и устранить проблемы в этих данных. 

Содержит информацию о сердечных болезнях.
1.	возраст
2.	пол
3.	тип боли в груди (4 значения [TA: типичная стенокардия, ATA: атипичная стенокардия, NAP: неангинальная боль, ASY: бессимптомная])
4.	артериальное давление в покое
5.	холестерин сыворотки в мг/дл
6.	уровень сахара в крови натощак > 120 мг/дл
7.	ЭКГ в покое: результаты электрокардиограммы в покое норма: нормальная, ST: аномалия ST-T LVN: гипертрофия
8.	MaxHR: максимальная достигнутая частота сердечных сокращений [Числовое значение от 60 до 202]
9.	стенокардия, вызванная физической нагрузкой (да, нет)
10.	 oldpeak = депрессия ST, вызванная физической нагрузкой, по сравнению с состоянием покоя
11.	наклон пикового сегмента ST при нагрузке (Вверх: восходящий, Плоский: плоский, Вниз: нисходящий)
12.	 HeartDisease: выходной класс [1: болезнь сердца, 0: нормальный]


Загружаю датасет с помощью библиотеки Pandas и вывожу первые 20 строк.

In [10]:
import pandas as pd

heart_df = pd.read_csv('3heart.csv', sep=',')
pd.options.display.expand_frame_repr = False
print(heart_df.head(20))

    Age Sex ChestPainType  RestingBP Cholesterol  FastingBS RestingECG  MaxHR ExerciseAngina  Oldpeak ST_Slope  HeartDisease
0    40   M           ATA      140.0         289          0     Normal  172.0              N      0.0       Up             0
1    49   F           NAP      160.0         180          0     Normal  156.0              N      1.0     Flat             1
2    37   M           ATA      130.0         283          0         ST   98.0              N      0.0       Up             0
3    48   F           ASY      138.0         214          0     Normal  108.0              Y      1.5     Flat             1
4    54   M           NAP      150.0         195          0     Normal  122.0              N      0.0       Up             0
5    39   M           NAP      120.0         339          0     Normal  170.0              N      0.0       Up             0
6    45   F           ATA      130.0         237          0     Normal  170.0              N      0.0       Up             0


Цель датасета увидеть зависимость наличия у людей болезни сердца от разных жизненных показательней: Возраста; Пола; Типа боли в груди; Артериального давления в покое; Холестерина сыворотки; Уровень сахара в крови натощак; ЭКГ в покое; Максимального сердечного ритма; Стенокардии; Старого пика; Наклона пикового сегмента ST при нагрузке.

С помощью методов .info и. columns попытаюсь оценить данные.

In [6]:
print(heart_df.info(), '\n')


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 925 entries, 0 to 924
Data columns (total 12 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Age             925 non-null    int64  
 1   Sex             925 non-null    object 
 2   ChestPainType   925 non-null    object 
 3   RestingBP       923 non-null    float64
 4   Cholesterol     924 non-null    object 
 5   FastingBS       925 non-null    int64  
 6   RestingECG      924 non-null    object 
 7   MaxHR           924 non-null    float64
 8   ExerciseAngina  925 non-null    object 
 9   Oldpeak         925 non-null    float64
 10  ST_Slope        925 non-null    object 
 11  HeartDisease    925 non-null    int64  
dtypes: float64(3), int64(3), object(6)
memory usage: 86.8+ KB
None 



In [7]:
print(heart_df.columns, '\n')

Index(['Age', 'Sex', 'ChestPainType', 'RestingBP', 'Cholesterol', 'FastingBS',
       'RestingECG', 'MaxHR', 'ExerciseAngina', 'Oldpeak', 'ST_Slope',
       'HeartDisease'],
      dtype='object') 



Как видно из первых строк датафрейм состоит из 925 строк и 12 столбцов. Названия столбцов с точки зрения логики и грамматики соответствуют правильным. Так же из данных столбца Non-Null Count можно сделать вывод что 3,4,6,7 столбце датафрейма присутствуют некорректные значения. Типы данных для столбцов не требуют измениения, кроме Cholesterol, требуется заменить object на int64

Начал процесс формализации данных с удаления явных дубликатов посредством метода .drop_duplicates() и восстановил индексирование методом .reset_index(). Результат работы метода на рисунке 3.

In [11]:
print('Кол-во дубликатов: ', heart_df.duplicated().sum(), '\n')
print('Индексы строк дубликатов:')
for index, value in enumerate(heart_df.duplicated()):
    if value:
        print(index)
heart_df = heart_df.drop_duplicates()
print('Удаление дубликатов заершено', '\n')
print('Кол-во дубликатов: ', heart_df.duplicated().sum(), '\n')

Кол-во дубликатов:  3 

Индексы строк дубликатов:
918
919
920
Удаление дубликатов заершено 

Кол-во дубликатов:  0 



После избавился от строк содержащих Na с помощью метода .dropna() и восстановил индексирование методом .reset_index().

In [12]:
print("Столбцы содержащие Na", '\n')
print(heart_df.isna().sum(), '\n')
heart_df = heart_df.dropna()
print("Удаление строк с Na завершено", '\n')
print(heart_df.isna().sum(), '\n')

heart_df.reset_index(drop=True, inplace=True)

Столбцы содержащие Na 

Age               0
Sex               0
ChestPainType     0
RestingBP         2
Cholesterol       1
FastingBS         0
RestingECG        1
MaxHR             1
ExerciseAngina    0
Oldpeak           0
ST_Slope          0
HeartDisease      0
dtype: int64 

Удаление строк с Na завершено 

Age               0
Sex               0
ChestPainType     0
RestingBP         0
Cholesterol       0
FastingBS         0
RestingECG        0
MaxHR             0
ExerciseAngina    0
Oldpeak           0
ST_Slope          0
HeartDisease      0
dtype: int64 



Поиск неявных дубликатов провел методом .unique() примененным к каждому столбцу. Ошибки были найдены в столбцах: Sex, значение может быть только M или F; RestingBP, не может быть 0 значения; Cholesterol, возможно только число; ST_Slope, значение может быть только Up, Down или Flat.

Исправлю некорректно написанные значения методом .replace().

In [13]:
for column in heart_df.columns:
    print('Уникальные значения поля', column, ': ', heart_df[column].unique())

heart_df["Sex"] = heart_df["Sex"].replace('Ma', 'M')
heart_df["RestingBP"] = heart_df["RestingBP"].replace('0', '100')
heart_df["Cholesterol"] = heart_df["Cholesterol"].replace('0', '200')
heart_df["Cholesterol"] = heart_df["Cholesterol"].replace('a241', '241')
heart_df["ST_Slope"] = heart_df["ST_Slope"].replace('UP', 'Up')
heart_df['Cholesterol'] = heart_df['Cholesterol'].astype('int64')

print('\n', "Редактироване ошибок завершено", '\n')

Уникальные значения поля Age :  [40 49 37 48 54 39 45 58 42 38 60 36 43 44 53 52 51 56 41 32 65 35 59 50
 47 31 46 57 55 63 66 34 33 61 29 62 28 30 74 68 72 64 69 67 73 70 77 75
 76 71]
Уникальные значения поля Sex :  ['M' 'F' 'Ma']
Уникальные значения поля ChestPainType :  ['ATA' 'NAP' 'ASY' 'TA']
Уникальные значения поля RestingBP :  [140. 160. 130. 138. 150. 120. 110. 136. 115. 100. 124. 113. 125. 145.
 112. 132. 118. 170. 142. 190. 135. 180. 108. 155. 128. 106.  92. 200.
 122.  98. 105. 133.  95.  80. 137. 185. 165. 126. 152. 116.   0. 144.
 154. 134. 104. 139. 131. 141. 178. 146. 158. 123. 102.  96. 143. 172.
 156. 114. 127. 101. 174.  94. 148. 117. 192. 129. 164.]
Уникальные значения поля Cholesterol :  ['289' '180' '283' '214' '195' '339' '237' '208' '207' '284' '211' '164'
 '204' '234' '273' '196' '248' '267' '223' '184' '201' '288' '209' '260'
 '468' '188' '518' '167' '224' '172' '186' '254' '306' '250' '177' '227'
 '230' '294' '264' '259' '175' '318' '216' '340' '233' '205' '

Составлю сводную таблицу методом pd.pivot_table() зависимости наклона пикового сегмента от среднего возраста и разделю значения по полу.

Напишу несколько группировок при помощи метода .groupby. Первая это группировка по полу и среднему артериальному давлению, вторая по типу боли и среднему холестерину, и артериальному давлению.

In [14]:
print(pd.pivot_table(heart_df, index='Sex', columns='ST_Slope', values='Age', aggfunc='mean'), '\n')

agg_func_mean = {'RestingBP': ['mean']}
print(heart_df.groupby(['Sex']).agg(agg_func_mean).round(2), '\n')

agg_func_mean = {'Cholesterol': ['mean'], 'RestingBP': ['mean']}
print(heart_df.groupby(['ChestPainType']).agg(agg_func_mean).round(2), '\n')
print('\n')

ST_Slope       Down       Flat         Up
Sex                                      
F         59.714286  54.618421  50.690909
M         58.142857  55.387013  50.802817 

    RestingBP
         mean
Sex          
F      132.32
M      132.45 

              Cholesterol RestingBP
                     mean      mean
ChestPainType                      
ASY                237.46    133.24
ATA                241.25    130.68
NAP                231.01    130.97
TA                 228.80    136.41 





<h3>Вывод</h3>


<pharagraph>
    Расширил познания в библиотеке Pandas. Научился редактировать данные, находить и удалять дубликаты, искать уникальные значения, с помощью которых возможна последующая формализация данных – осуществлять предварительную обработку данных.
Узнал о сводных таблицах библиотеки Pandas и методе группировки данных groupby. 
</pharagraph>